[Feature] Warning Screen for High Tx Fees

Chaitanya-Keyal opened the pull request:
## Description

This PR introduces a warning screen that appears before the PSBT math screen if the transaction includes unusually high fees. The goal is to ensure the user is explicitly made aware of any excessive fees before proceeding to sign the transaction.

Closes #721

## Warning Condition
The warning is shown when:
- The fee amount exceeds 25% of the sum of all outputs excluding change, and
- There are outputs other than _true_ change (and fee).

## New Warning Screen:

<img width="240" height="240" alt="image" src="https://github.com/user-attachments/assets/597d624d-e58f-4d47-b862-c67027a28f9c" />

## Visual Warning in PSBT Flow Views

<img width="240" height="240" alt="image" src="https://github.com/user-attachments/assets/e808aa61-3007-41e8-b435-c4fdee3ea299" /> <img width="240" height="240" alt="image" src="https://github.com/user-attachments/assets/d142b391-c77a-4c23-a969-9fcac0ace075" />

---

This pull request is categorized as a:

- [x] New feature

## Checklist

- [x] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [x] Yes

The warning screen doesn’t currently apply to any of the existing test PSBT flows. I also noticed there aren’t tests for similar warnings (e.g. Full Spend), so figured it's not needed here either. Do let me know if a test needs to be added!

I have tested this PR on the following platforms/os:

- [x] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)

Chaitanya-Keyal committed:
feat: add a warning for high tx fee

Chaitanya-Keyal committed:
feat: add a warning for high tx fee

Chaitanya-Keyal commented on l10n/messages.pot:
```diff
@@ -1254,11 +1265,11 @@ msgid "Review SeedQR"
 
 #: src/seedsigner/views/seed_views.py
-msgid "Your transcribed SeedQR successfully scanned and yielded the same seed."
+msgid "Your transcribed SeedQR could not be read!"
```
`python setup.py extract_messages` seems to have reordered 2 these strings (probably to be consistent with the order they are used in the code)

kdmukai commented on src/seedsigner/views/psbt_views.py:
```diff
@@ -151,10 +151,18 @@ def run(self):
+
+        # Huge Fee if fee amount > 25% of sum of all outputs excluding change
+        # Skip the warning when there is no output other than change
+        elif sum_of_all_outputs_excluding_change > 0 and psbt_parser.fee_amount > 0.25 * (sum_of_all_outputs_excluding_change):
```
Maybe add a `TODO` comment here that we might want to make this 25% threshold configurable some day.

Any hard-coded value like this just feels a bit wrong in general.

Chaitanya-Keyal replied:
Agreed — I actually mentioned this in the PR description as well. Making the threshold configurable definitely makes sense, though I think it’s fine to defer that for now. I’ve added a `TODO` comment and defined a named constant for the 25% value so that we can easily update it later without requiring translation changes to the warning message.

kdmukai commented on src/seedsigner/views/psbt_views.py:
```diff
@@ -204,6 +212,32 @@ def run(self):
+class PSBTHugeFeeWarningView(View):
+    def run(self):
+        selected_menu_num = WarningScreen(
+            status_headline=_("Huge Fee!"),
```
I know the original inspiration for this feature was a similar warning in a different hardware wallet. But I don't know that we need to / should use the same word choices.

"Huge" just strikes me as kind of a comical word choice.

"Very large"?

We can't really say "Excessive" because we'll never know the current fee market.

Also not sure how consistent we are with exclamation points in these kinds of status headlines. The screen itself has a giant exclamation point already.

@easyuxd: sorry to keep tagging you in on these, but want to make a definitive call about exclamation points here? I checked your https://github.com/easyuxd/seedsigner-design-system and didn't see any guidelines specifically for warning messages. See the PR description for a screenshot. My high school AP English teacher was pretty savage about exclamation points being unnecessary ("if you don't make the point forcefully with your words, adding something at the end isn't going to do the job for you" (refraining from ending that w/an exclamation point...). So I tend to lean anti-exclamation point in professional / product settings.

caveat: Exclamation point as iconography (i.e. the big exclamation point in the screen design itself) is a separate use case that I am completely happy with.

Chaitanya-Keyal replied:
Good point — I’m definitely open to alternative wording. I’ve updated the PR to use "High Fee" instead of "Huge Fee" across the board.

My original choice was mainly for consistency with the existing warning screens like "Unsupported Script Type!" and "Full Spend!", both of which also use exclamation points. That said, I’m happy to adjust further if we decide to adopt a different tone or style for warnings going forward.

kdmukai commented on src/seedsigner/views/psbt_views.py:
```diff
@@ -204,6 +212,32 @@ def run(self):
+    def run(self):
+        selected_menu_num = WarningScreen(
+            status_headline=_("Huge Fee!"),
+            text=_("This PSBT has a network fee higher than 25% of the total output value excluding change."),
```
Do we say "network fee" anywhere else? Pretty sure we only use the term "fee".

Also I would place "excluding change" within parentheses.

kdmukai replied:
If other contributors think it's likely that we might want to make the 25% threshold configurable some day, perhaps this string should use a variable insertion for that percentage so that this sentence doesn't have to be re-translated in the future.

Chaitanya-Keyal replied:
I’ve updated the text to say just "fee" for consistency. I also added (excluding change) in parentheses as suggested and updated the string to use a variable for the threshold percentage.

kdmukai commented on src/seedsigner/views/psbt_views.py:
```diff
@@ -151,10 +151,18 @@ def run(self):
 
+        sum_of_all_outputs = sum([psbt_parser.psbt.tx.vout[i].value for i in range(len(psbt_parser.psbt.outputs))])
+        sum_of_all_outputs_excluding_change = sum_of_all_outputs - psbt_parser.change_amount
+
```
This would probably be better as a method in `psbt_parser` so that a unit test could be written just for that isolated functionality.

Chaitanya-Keyal replied:
Done! I’ve moved the logic into its own method in psbt_parser, refactored it a bit to make things cleaner, and added a corresponding unit test.

kdmukai commented:
Ideally this PR would include a flow test for the scenario that triggers this new View. But unfortunately we are missing a flow test for other similar psbt warning screens (e.g. "full spend"), so we don't have an easy copy-paste flow test to start from.

As noted elsewhere, I DO think we should get a unit test on the total output value calcs.

So:
* refactor and add unit test: Blocker
* flow test: nice to have; not a blocker
* screenshot: Nice job already adding it!

Chaitanya-Keyal commented:
Hey @kdmukai,

Thanks a lot for the review!

I've updated the PR to incorporate all your suggestions:
- Added two unit tests for checking the output value calculation:
  - One using mocked PSBT data.
  - One using a real PSBT (reused from an existing test). I think the mocked test should be sufficient, but included both for completeness — let me know if you'd prefer to drop the second one.
- Updated the terminology from "huge fee" to "high fee" — noticed Sparrow Wallet also uses "high fee", and it does sound more natural. Open to renaming if you have a better suggestion.
- Added a `TODO` comment for possibly making the threshold configurable via settings. The threshold value is now a named constant, so translations won't have to be redone in the future.

As for the flow test for this screen (and other PSBT warning views like "full spend"), it probably makes more sense to tackle that in a follow-up PR rather than squeeze it into this one. I can take that up once this is merged.

kdmukai commented on src/seedsigner/models/psbt_parser.py:
```diff
@@ -390,3 +390,13 @@ def verify_multisig_output(self, descriptor: Descriptor, change_num: int) -> boo
+            and show a warning screen if it is.
+        """
+        total_outputs = sum(out.value for out in self.psbt.tx.vout)
+        return total_outputs - self.change_amount
```
I can't recall the details without digging deeper into the code, but is the `self.change_amount` smart about differentiating change from a self-transfer addr?

Also would it make sense to alter this method to be more like:
```python
def total_output_value(self, include_change: bool = True) -> int:
```

Though I wonder if there's already a similar method (again, just doing a quick review here so I'm being lazy and not digging into the code myself).

kdmukai replied:
(`get_sum_of` just seems a bit too clunky to me when `total` says the same thing. And I think we tend to avoid `get` naming convention as those sorts of calls are usually declared as `@property` and such, by convention, you omit the `get`)

Chaitanya-Keyal replied:
> I can't recall the details without digging deeper into the code, but is the self.change_amount smart about differentiating change from a self-transfer addr?

Good catch! `self.change_amount` actually does not differentiate between change and self-transfer on its own. The `PSBTChangeDetailsView` uses the derivation path (specifically the chain index) to determine whether an output is a self-transfer or true change, and then renders the appropriate screen.
I've updated this PR to use essentially the same logic — so this function now only subtracts *true* change (i.e., change that is not a self-transfer).

> I wonder if there's already a similar method

I don't think such a function exists, since this use case hadn’t come up before. Let me know if I’ve missed anything.

> tend to avoid get naming convention as those sorts of calls are usually declared as @property

Totally makes sense in the context of property-style methods. Since this one takes a parameter (`include_change`), it can’t be a `@property`, so I kept the `get_` prefix for clarity. I also followed the naming precedent from `get_change_data()`, which is similarly structured and not a property either.

That said, I’m open to renaming it to `total_output_value()` if you feel that reads better. Also — if there’s a strong reason to support the `include_change` param now, happy to keep it, but at the moment I’m not sure there’s a concrete use case where we’d actually want to pass `True`. Please correct me if I’m missing something there.

kdmukai commented on src/seedsigner/models/psbt_parser.py:
```diff
@@ -390,3 +390,13 @@ def verify_multisig_output(self, descriptor: Descriptor, change_num: int) -> boo
+    
+    def get_sum_of_outputs_excluding_change(self):
+        """
+            Returns the sum of all outputs excluding change amount.
```
It's implicit in this description, but wouldn't hurt to explicitly note that the fee is excluded from this result.

Chaitanya-Keyal commented:
Hey @kdmukai,

Thanks again for your quick review! I've updated the PR with your suggestions:
- Fixed the function not including self-transfers as output
- Updated tests and added one more case to verify self transfer + change calculations

While I was working on these updates, I realized that it could be helpful to introduce a new variable in `PSBTParser` to explicitly track self-transfers. This would simplify calculations like the one in this function and also improve the screen selection logic in `PSBTChangeDetailsView`. By doing so, we can more easily differentiate between self-transfers and true change outputs, making the logic clearer and more reusable across the codebase.

Let me know your thoughts on this approach! If it sounds good, perhaps I'll start to work on that on another branch!

kdmukai commented on tests/test_psbt_parser.py:
```diff
@@ -201,3 +203,72 @@ def test_parse_op_return_content():
+    
+    # Expected: only the destination output (319,049,328 sats)
+    expected = 319049328
+    assert pp.get_total_output_value() == expected
```
If the `include_change` option is going to remain in the function, should also test with that option on.

kdmukai replied:
Can be in the same test. Just an extra pair of `expected` and `assert`. Though maybe update the test name then. And the test description comment.

Chaitanya-Keyal replied:
That makes sense, I've updated the test to reflect these changes.

newtonick commented:
tACK. I reviewed the code but would love to get some additional eyes (beyond Keith and I) on this one since its code changes around parsing the PSBT amounts and displaying them to the user.

I've also not been consistent about the PR expectation should be around including an updated messages.pot files when new text is included that needs translation. So when/if this PR gets merged it would be knowing that the messages.pot would still need to be regenerated/extracted for the dev branch.

alvroble commented:
ACK tested as of https://github.com/SeedSigner/seedsigner/pull/722/commits/8910071dfc0de821e09cd0fef38d241f1ec2b423 on Raspberry Pi Manual OS. 

You could take it one step further. Sparrow displays a visual warning for the fee in the PSBT overview. I propose using the same approach here, perhaps by coloring the fees red when they are too high in the `PSBTOverviewView` and/or `PSBTMathView` (in addition to the warning view). This is because, when the user navigates back and forth, the `PSBTHighFeeWarningView` no longer appears.

The UI would look like this for the `PSBTOverviewView` (and something similar for `PSBTMathView`):

<img width="240" height="240" alt="imagen" src="https://github.com/user-attachments/assets/fce1ffeb-ef39-448a-93c1-7d2e649f5cc3" />

Maybe even adding some kind of warning sing like `(!)` or `(high!)` next to fee would be also useful too.

Feel free to check or cherry-pick the modifications from my repo https://github.com/alvroble/seedsigner/commit/4f01eec88b7e8a0dcd608df62b2bb8ef63c08a31.

Chaitanya-Keyal committed:
feat: highlight high fees in PSBTOverview and PSBTMath Screen

Co-authored-by: alvroble <50918598+alvroble@users.noreply.github.com>

Chaitanya-Keyal committed:
feat: highlight high fees in PSBTOverview and PSBTMath Screen

Co-authored-by: alvroble <50918598+alvroble@users.noreply.github.com>

Chaitanya-Keyal commented:
Thanks @alvroble! That’s a great idea. I’ve updated the PR accordingly and added you as a co-author on the commit.

If the fee is high, both `PSBTOverviewScreen` and `PSBTMathScreen` now display "fee (!)" in the dire warning color.

I also moved the high-fee check into its own function in `PSBTParser` and added a corresponding unit test.

alvroble commented:
Thanks @Chaitanya-Keyal!! Now it looks like this:

PSBTOverviewView
<img width="240" height="240" alt="imagen" src="https://github.com/user-attachments/assets/712993e3-778a-4ef3-a50d-d50f8f834274" />

PSBTMathView
<img width="240" height="240" alt="imagen" src="https://github.com/user-attachments/assets/1d3d862b-0243-45db-8ae6-7d80acfed2a1" />

Chaitanya-Keyal commented:
I've updated the tests following the addition of the new, exhaustive `TestPSBTParser` class. The `get_total_output_value()` method will now be tested for each PSBT in the list, and all other parameterized and specific tests have also been moved inside the class.

Also, I’ve reworded the warning to use "transaction" instead of "PSBT", in line with the recent changes that do the same. PR description has the updated screenshots.

newtonick commented:
@easyuxd would you review the change in this PR? I'm interested in your thoughts on adding a new warning screen for a high transaction fee and the visual indications around having a high fee.

easyuxd commented:
Thanks for tagging me, @newtonick. 

I think this warning screen is a useful UX improvement, @Chaitanya-Keyal. I would recommend changing the message type from “Warning” to “Dire Warning.” This was a recent addition, I new message type I introduced that is more severe than a warning but not an error, and it encompasses this “high-attention / loss of funds” use case:

https://github.com/easyuxd/seedsigner-ux-guidelines?tab=readme-ov-file#dire-warning

I would also recommend changing the “fee(!)” label color to this DIRE_WARNING_COLOR to match.

Chaitanya-Keyal commented:
Thanks @easyuxd! I've updated the view to use `DireWarningScreen` and the high fee label was already using `DIRE_WARNING_COLOR`. Updated screenshot is in the description, lmk what you think!

easyuxd commented:
@Chaitanya-Keyal Thank you, looks great! ACK

--
Off-topic +@kdmukai:
This PR reminds me of how non-descriptive our titles are. It's not you, but how we originally designed these messaging screens. Since we focus on the screen title more than the small colored text below it, I had proposed removing the colored text and moving that content to the navbar. This would make the screens more scannable and free up valuable pixels. "Caution" doesn't mean much, especially when there's already a colored warning icon. Anyway, this type of change would be global, affecting all messaging screens and translations.

https://github.com/SeedSigner/seedsigner/issues/606

Given the current experience, "Caution" seems as consistent a title as any.
--
